bootstrap/bootstrap.go

```diff
@@ -231,116 +238,54 @@ func (b *Bootstrap) extractTraceCtx() opentracing.SpanContext {
-	jobURL := fmt.Sprintf("%s#%s", buildURL, b.JobID)
-	source, _ := b.shell.Env.Get("BUILDKITE_SOURCE")
 	label, hasLabel := b.shell.Env.Get("BUILDKITE_LABEL")
 	if !hasLabel {
 		label = "job"
 	}
-	retry := int64(0)
-	if attemptStr, has := b.shell.Env.Get("BUILDKITE_RETRY_COUNT"); has {
-		if parsedRetry, err := strconv.ParseInt(attemptStr, 10, 64); err == nil {
-			retry = parsedRetry
-		}
-	}
-	parallel := int64(0)
-	if parallelStr, has := b.shell.Env.Get("BUILDKITE_PARALLEL_JOB"); has {
-		if parsedParallel, err := strconv.ParseInt(parallelStr, 10, 64); err == nil {
-			parallel = parsedParallel
-		}
-	}
-	rebuiltFromID, has := b.shell.Env.Get("BUILDKITE_REBUILT_FROM_BUILD_NUMBER")
-	if !has || rebuiltFromID == "" {
-		rebuiltFromID = "n/a"
-	}
-	triggeredFromID, has := b.shell.Env.Get("BUILDKITE_TRIGGERED_FROM_BUILD_ID")
-	if !has || triggeredFromID == "" {
-		triggeredFromID = "n/a"
+
+	resourceName := b.OrganizationSlug + "/" + b.PipelineSlug + "/" + label
+	opts := []tracer.StartOption{
+		tracer.WithServiceName("buildkite_agent"),
+		tracer.WithSampler(tracer.NewAllSampler()),
+		tracer.WithAnalytics(true),
+		tracer.WithGlobalTag(ddext.ResourceName, resourceName),
 	}
 
-	// Set specific tracing library here. Everything else should be using opentracing.
-	// Use a constant sampler - CI runs aren't high traffic.
-	var t opentracing.Tracer
-	var stopper stopper
-	switch b.Config.TracingBackend {
-	case "datadog":
-		t = opentracer.New(
-			tracer.WithServiceName("buildkite_agent"),
-			tracer.WithSampler(tracer.NewAllSampler()),
-			tracer.WithAnalytics(true),
-			tracer.WithGlobalTag("buildkite.agent", b.AgentName),
-			tracer.WithGlobalTag("buildkite.version", agent.Version()),
-			tracer.WithGlobalTag("buildkite.queue", b.Queue),
-			tracer.WithGlobalTag("buildkite.org", b.OrganizationSlug),
-			tracer.WithGlobalTag("buildkite.pipeline", b.PipelineSlug),
-			tracer.WithGlobalTag("buildkite.branch", b.Branch),
-			tracer.WithGlobalTag("buildkite.job_id", b.JobID),
-			tracer.WithGlobalTag("buildkite.job_url", jobURL),
-			tracer.WithGlobalTag("buildkite.build_id", buildID),
-			tracer.WithGlobalTag("buildkite.build_number", buildNumber),
-			tracer.WithGlobalTag("buildkite.build_url", buildURL),
-			tracer.WithGlobalTag("buildkite.source", source),
-			tracer.WithGlobalTag("buildkite.retry", retry),
-			tracer.WithGlobalTag("buildkite.parallel", parallel),
-			tracer.WithGlobalTag("buildkite.rebuilt_from_id", rebuiltFromID),
-			tracer.WithGlobalTag("buildkite.triggered_from_id", triggeredFromID),
-			tracer.WithGlobalTag(ddext.SamplingPriority, ddext.PriorityUserKeep),
-		)
-		stopper = tracer.Stop
-	default:
-		b.shell.Commentf("An invalid tracing backend was given: %s. Tracing will not occur.", b.Config.TracingBackend)
-		fallthrough
-	case "":
-		t = opentracing.NoopTracer{}
-		stopper = func() {}
+	tags := Merge(GenericTracingExtras(b, *b.shell.Env), DDTracingExtras())
+	opts = slices.Grow(opts, len(tags))
```

Comment:
review note: i'm using the new in Go 1.18, experimental [`slices`](https://pkg.go.dev/golang.org/x/exp/slices) package to grow our slice to the correct size here.

it's experimental in that the API might change, rather than experimental in that it might not work (like this PR lol), but regardless, i'm happy to replace this with non-generic code if we're not comfortable using it.

Reply:
if you're wondering why i manually grow the slice rather than just appending willy nilly, when you call append on a slice in go, if there isn't enough capacity in the slice to add the element you want to add, it automatically doubles the capacity of the slice. This is fine in most cases (and honestly, probably fine here), but it can mean that we allocate more memory for a given slice than we strictly speaking need. 

Because we know how big the slice is going to get ahead of time (or at least, we can derive it from the sizes of other things) it pays to allocate only as much memory to the slice.